scalyr_agent/agent_main.py

```diff
@@ -152,6 +155,9 @@ def __init__(self, platform_controller):
         # possible.
         self.__run_state = None
 
+        # Store references to the last value of the OverallStats instance
+        self.__overall_stats = OverallStats()
```

Comment:
@czerwingithub I was a bit confused by the ``base_overall_stats``variable so maybe you can clarify it. It's likely though that I'm misunderstanding the comment / how it should work - https://github.com/scalyr/scalyr-agent-2/blob/ab68a8f68be646dd7c43cc6dc8f473d440ee78f8/scalyr_agent/agent_main.py#L849.

It says it tracks "the accumulated stats since the last stat reset", but that doesn't appear to be the case. That value is only ever updated when config is reloaded.

In a place where we re-calculate overall stats we don't update that variable - https://github.com/scalyr/scalyr-agent-2/blob/master/scalyr_agent/agent_main.py#L956.

And inside, ``__calculate_overall_stats`` method we also don't manipulate this variable - we return a new overall stats object (https://github.com/scalyr/scalyr-agent-2/blob/master/scalyr_agent/agent_main.py#L1516).

According to the comment and my understanding, for that variable to "comply" with the documented behavior, we should re-assign it every time we re-calculate the stats (aka here https://github.com/scalyr/scalyr-agent-2/blob/master/scalyr_agent/agent_main.py#L957).

So either I'm misunderstanding something (likely) or this is not working as it should / as documented in the comment.

---

That's also the reason why I introduced a new variable which basically does what, according to the my understanding of that comment, ``base_overall_stats`` variable should be doing (I wasn't too sure about the existing code behavior and I didn't want to change it).

Reply:
Ok, that first comment is probably not a good explanation for what's going on.

The `base_overall_stats` is meant to capture the value of all stats when we did the last configuration update.  It is used to properly calculate the value of the stats since the start of the agent.

Here's the problem it's trying to solve.  Whenever we do a configuration update, we do a soft restart of the agent, creating new instances of objects like the CopyingManager.  The CopyingManager only has the ability to tell us how many bytes it is has copied since it was created... which of course means we only have the number of bytes copied since the last configuration change.  That's not sufficient for what we'd like to do.  We want to report how many bytes have been copied since the start of the agent, no matter how many different CopyingManagers we've used over time.

That's where the `base_overall_stats` comes in.  It holds the number of bytes the agent process has copied before the last reset.  So, to get a complete number (the number of bytes since the agent process has started), we just add in the value in `base_overall_stats` (representing copied bytes before last reset) to the value recorded by the current CopyingManager (representing copied bytes since last reset).  That's also why we only update `base_overall_stats` on configuration resets.  We are about to lose the current CopyingManager, so we need to add in its contribution to the `base_overall_stats` to essentially create a new baseline... we will then add to the new CopyingManager's stats.

Does that make sense?

Reply:
Thanks for the clarification. This makes sense, yes.

It would be good to incorporate some of that clarification in the comment itself :)